XmlFileType.js

```diff
@@ -0,0 +1,541 @@
+        if (pseudo) {
+            this.pseudos[locale] = pseudo;
+        }
+    }.bind(this));
+
+    // for use with missing strings
+    if (!project.settings.nopseudo) {
+        this.missingPseudo = this.API.getPseudoBundle(project.pseudoLocale, this, project);
+    }
+
+    this.schemas = {};
+    this.refs = {};
+    this.loadSchemas(".");
+};
+
+/** join two paths unless the child is an absolute path */
+function smartJoin(parent, child) {
+    if (!parent) return child;
+    if (!child) return parent;
+    return (child[0] === "/") ? child : path.join(parent, child);
+}
+
+XmlFileType.prototype.loadSchemaFile = function(pathName) {
+    // only read json files as schemas
+    if (!pathName.endsWith(".json") && !pathName.endsWith(".jsn")) return;
+
+    try {
+        var schema = fs.readFileSync(pathName, "utf-8");
+        var schemaObj = JSON.parse(schema);
+        this.schemas[pathName] = schemaObj;
+        this.refs[schemaObj["$id"]] = schemaObj;
+    } catch (e) {
+        // logger.fatal("Error while parsing schema file " + pathName);
+        console.log("Error while parsing schema file " + pathName);
+        throw e;
+    }
+};
+
+XmlFileType.prototype.loadSchemaDir = function(pathName) {
+    var files = fs.readdirSync(pathName);
+    if (files) {
+        files.forEach(function(file) {
+            var full = smartJoin(pathName, file);
+            this.loadSchemaDirOrFile(full);
+        }.bind(this));
+    }
+};
+
+XmlFileType.prototype.loadSchemaDirOrFile = function(pathName) {
+    var stats = fs.statSync(pathName);
+    if (!stats) return;
+    if (stats.isDirectory()) {
+       this.loadSchemaDir(pathName);
+    } else {
+       this.loadSchemaFile(pathName);
+    }
+};
+
+
+XmlFileType.prototype.findRefs = function(root, schema, ref) {
+    if (typeof(schema) !== 'object') return;
+
+    if (typeof(root["$$refs"]) === 'undefined') {
+        root["$$refs"] = {};
+    }
+
+    root["$$refs"][ref] = schema;
+
+    // currently for simplicity, we only handle intra-schema
+    // references so far. We'll have to implement the extra-schema
+    // references later
+    for (var prop in schema) {
+        var subschema = schema[prop];
+
+        // recurse through everything -- if it is not a
+        // schema object, that's okay -- it won't have the
+        // $anchor in it
+        if (prop === '$anchor') {
+            root["$$refs"]["#" + subschema] = schema;
+        } else if (prop !== '$$refs') {
+            var newref = ref + '/' + prop;
+            if (Array.isArray(subschema)) {
+                subschema.forEach(function(element, index) {
+                    var arrayref = newref + '[' + index + ']';
+                    this.findRefs(element, arrayref);
+                }.bind(this));
+            } else if (typeof(subschema) === 'object') {
+                this.findRefs(root, subschema, newref);
+            }
+        }
+    }
+};
+
+var defaultSchema = {
```

Comment:
I think it makes sense to extract this to a separate file and load it from there, in order to make easier to use it as an example for users of the plugin.

Reply:
put in a separate file